momaland/envs/gem_mining/gem_mining.py

```diff
@@ -0,0 +1,274 @@
+        self.episode_num = 0
+
+        infos = {agent: {} for agent in self.agents}
+        return observations, infos
+
+    def step(self, actions):
+        """Steps in the environment.
+
+        Args:
+            actions: a dict of actions, keyed by agent names
+
+        Returns: a tuple containing the following items in order:
+        - observations
+        - rewards
+        - terminations
+        - truncations
+        - infos
+        dicts where each dict looks like {agent_1: item_1, agent_2: item_2}
+        """
+        # If a user passes in actions with no agents, then just return empty observations, etc.
+        if not actions:
+            self.agents = []
+            return {}, {}, {}, {}, {}
+
+        # - Observations -#
+        # return constant observations '0' as this is a stateless setting
+        observations = {agent: np.array([0], dtype=np.float32) for agent in self.agents}
+
+        # - Rewards -#
+        # First, calculate the number of workers ending up at each mine:
+        workers_at_mine = [0] * self.num_mines
+        for agent, mine in actions.items():
+            workers_at_mine[mine] = workers_at_mine[mine] + self.workers[agent]
+        print(workers_at_mine)
+        # The rewards are based on Bernoulli (binomial(1)) experiments per mine per objective
+        reward_vec = np.array([0] * self.num_objectives, dtype=np.float32)
+        for i in range(self.num_mines):
+            if workers_at_mine[i] > 0:
+                bonus = pow(self.worker_bonus, workers_at_mine[i])
+                for j in range(self.num_objectives):
+                    prob = bonus * self.base_probabilities[i][j]
+                    if prob > self.truncation_probability:
+                        prob = self.truncation_probability
+                    outcome = self.np_random.binomial(size=1, n=1, p=prob)
+                    reward_vec[j] = reward_vec[j] + outcome[0]
+        # every agent gets the same reward vector (fully cooperative)
+        rewards = dict()
+        for i in range(len(self.agents)):
+            rewards[i] = reward_vec
+
+        # - Infos -#
+        # typically there won't be any information in the infos, but there must still be an entry for each agent
+        infos = {agent: {} for agent in self.agents}
+
+        # stateless bandit setting where each episode only lasts 1 timestep
+        self.terminations = {agent: True for i, agent in enumerate(self.agents)}
+        self.agents = []
+
+        if self.render_mode == "human":
+            self.render()
+
+        return observations, rewards, self.truncations, self.terminations, infos
+
+    # - Helper Methods (prefix by _) -#
+    def _random_action(self):
+        """Draw a valid random action for the MOGemMining instance.
+
+        Returns:
+            A valid random joint action
+        """
+        result = dict()
+        for i in self.agents:
+            result[i] = self.action_spaces[i].sample()
+        return result
```

Comment:
Again, this can be done with a dict comprehension. It is faster, uses less memory and is more "pythonic".
```Python
result = {i: self.action_spaces[i].sample() for i in range(self.agents)}
```